qudipy/qutils/matrices.py

```diff
@@ -1,315 +1,443 @@
+
+    '''
+
+    ##### Operator Object Manipulation #####
+
+    def __init__(self, operators=None, filename=None):
+        '''
+
+        Keyword Arguments
+        -----------------
+        operators: Dictionary
+            Dictionary of unitary operators to be used to initialize an object.
+            If left empty then the dictionary is going to be built from scratch
+            or loaded from an operator object file from the working directory.
+        filename: string
+            String that indicates the name of the existing object file which
+            contains a library of unitary operators.
+
+        Returns
+        -------
+        None.
+
+        '''
+        # Ensure initial operator dictionary only contains unitary items with
+        # the correct data structure, then assign to object
+        if filename is None:
+            if operators is None:
+                # Empty dictionary
+                self.operators = {}
+            elif operators is not None:
+                # Use user defined dictionary without existing object file
+                self.check_ops(operators)
+                self.operators = operators
+        elif  filename is not None:
+            
+            # Append '.npz' if only the filename with out file type suffix
+            if filename[-4:] != '.npz':
+                filename = f'{filename}.npz'
+
+            if operators is None:
+                # Load dictionary object file with no user defined dictionary
+                self.operators = self.load_ops(filename)
+                self.check_ops(self.operators)
+            elif operators is not None:
+                # Use user defined dictionary and append existing object file
+                self.operators = operators
+                loaded_ops = self.load_ops(filename)
+                self.add_operators(loaded_ops)
+                self.check_ops(self.operators)
+
+        # Default operator dictionary
+        default_ops = {
+                'PAULI_X': np.array([[0, 1], [1, 0]], dtype=complex),
+                'PAULI_Y': np.array([[0,-1.0j], 
+                    [1.0j, 0]],dtype=complex),
+                'PAULI_Z': np.array([[1, 0], [0, -1]], dtype=complex),
+                'PAULI_I': np.array([[1, 0], [0, 1]], dtype=complex)
+            }    
         
-    Keyword Arguments
-    -----------------
-    None.
-
-    Returns
-    -------
-    y_k: 2D complex array
-        Matrix Y_k of dimensions 2**N x 2**N 
-
-    """
-    if k == 1:
-        return np.kron(PAULI_Y, np.eye(2 ** (N - 1)))
-    y_k = PAULI_I
-    for m in range(2, N + 1):
-        if k is m:
-            y_k = np.kron(y_k, PAULI_Y)
-        else:
-            y_k = np.kron(y_k, PAULI_I)
-    return y_k
-
+        # Always have Pauli operators and the identity added. 
+        # NOTE: this will overwrite dictionary values with same key name if 
+        # loading a dictionary from an object file
+        self.add_operators(default_ops)
 
```

Comment:
You can overload the bracket operators to simplify referencing the dictionary even more:
`self.operators['PAULI_I'] --> self['PAULI_I']`

To do that, add the following code but double-check it works as it should (adapted from https://stackoverflow.com/questions/1957780/how-to-override-the-operator-in-python):

```suggestion
    def __getitem__(self, key):
        return self.operators['key']    # you can also handle a KeyError here if you want
        
    def __setitem__(self, key, val):
        self.add_operators({key : val})
        
    def __delitem__(self, key):
        del self.operators['key']
```

After you ensure it works, please simplify the code of this type everywhere: `self.operators['PAULI_I'] --> self['PAULI_I']`